File: app/src/main/java/org/xdty/callerinfo/presenter/MainBottomPresenter.java
package org.xdty.callerinfo.presenter;

import org.xdty.callerinfo.application.Application;
import org.xdty.callerinfo.contract.MainBottomContact;
import org.xdty.callerinfo.data.CallerDataSource;
import org.xdty.callerinfo.model.database.Database;
import org.xdty.callerinfo.model.db.Caller;
import org.xdty.callerinfo.model.db.InCall;
import org.xdty.callerinfo.model.db.MarkedRecord;
import org.xdty.callerinfo.model.setting.Setting;
import org.xdty.callerinfo.utils.Alarm;
import org.xdty.callerinfo.utils.Utils;

import javax.inject.Inject;

public class MainBottomPresenter implements MainBottomContact.Presenter {

    @Inject
    Setting mSetting;
    @Inject
    CallerDataSource mCallerDataSource;
    @Inject
    Database mDatabase;
    @Inject
    Alarm mAlarm;

    private InCall mInCall;
    private Caller mCaller;

    private MainBottomContact.View mView;

    public MainBottomPresenter(MainBottomContact.View view) {
        mView = view;
        Application.getAppComponent().inject(this);
    }

    @Override
    public void start() {
        mView.init(mInCall, mCaller);
    }

    @Override
    public void bindData(InCall inCall) {
        mInCall = inCall;
        mCaller = mCallerDataSource.getCallerFromCache(inCall.getNumber());
    }

    @Override
    public boolean canMark() {
        return mCaller.isMark() || mCaller.canMark() && mInCall.getDuration() > 0;
    }

    @Override
    public void markClicked(int viewId) {
        MarkedRecord.MarkType type = MarkedRecord.MarkType.fromResourceId(viewId);

        if (type != MarkedRecord.MarkType.CUSTOM) {
            String typeText = Utils.typeFromId(type.toInt());
            mCallerDataSource.updateCaller(mInCall.getNumber(), type.toInt(), typeText);
            mView.updateMarkName(typeText);
        }

        mView.updateMark(viewId, mCaller);
    }

    @Override
    public void markCustom(String text) {
        mCallerDataSource.updateCaller(mInCall.getNumber(), MarkedRecord.MarkType.CUSTOM.toInt(),
                text);
        mView.updateMarkName(text);
    }
}

File: app/src/main/java/org/xdty/callerinfo/presenter/MainBottomPresenter.kt
package org.xdty.callerinfo.presenter

import android.util.Log
import org.xdty.callerinfo.application.Application.Companion.appComponent
import org.xdty.callerinfo.contract.MainBottomContact.Presenter
import org.xdty.callerinfo.contract.MainBottomContact.View
import org.xdty.callerinfo.data.CallerDataSource
import org.xdty.callerinfo.model.MarkType
import org.xdty.callerinfo.model.database.Database
import org.xdty.callerinfo.model.db.Caller
import org.xdty.callerinfo.model.db.InCall
import org.xdty.callerinfo.model.setting.Setting
import org.xdty.callerinfo.utils.Alarm
import org.xdty.callerinfo.utils.Utils
import javax.inject.Inject

class MainBottomPresenter(private val mView: View) : Presenter {
    @Inject
    internal lateinit var mSetting: Setting
    @Inject
    internal lateinit var mCallerDataSource: CallerDataSource
    @Inject
    internal lateinit var mDatabase: Database
    @Inject
    internal lateinit var mAlarm: Alarm

    private var mInCall: InCall? = null
    private var mCaller: Caller? = null
    override fun start() {
        if (mInCall != null) {
            mView.init(mInCall!!, mCaller!!)
        } else {
            Log.e(TAG, "mInCall is null")
        }
    }

    override fun bindData(inCall: InCall?) {
        mInCall = inCall
        mCaller = mCallerDataSource.getCallerFromCache(inCall!!.number)
    }

    override fun canMark(): Boolean {
        return mCaller!!.isMark || mCaller!!.canMark() && mInCall!!.duration > 0
    }

    override fun markClicked(viewId: Int) {
        val type: MarkType = MarkType.fromResourceId(viewId)
        if (type != MarkType.CUSTOM) {
            val typeText = Utils.typeFromId(type.toInt())
            mCallerDataSource.updateCaller(mInCall!!.number, type.toInt(), typeText)
            mView.updateMarkName(typeText)
        }
        mView.updateMark(viewId, mCaller!!)
    }

    override fun markCustom(text: String?) {
        mCallerDataSource.updateCaller(mInCall!!.number, MarkType.CUSTOM.toInt(),
                text!!)
        mView.updateMarkName(text)
    }

    companion object {
        private val TAG = MainBottomPresenter::class.java.simpleName
    }

    init {
        appComponent.inject(this)
    }
}